<a href="https://colab.research.google.com/github/RobDBennett/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [0]:
### YOUR CODE STARTS HERE
import pandas as pd
import numpy as np
from scipy import stats
import math
import sys



In [0]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-05-12 15:09:34--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-05-12 15:09:35 (128 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']
df = pd.read_csv('house-votes-84.data', names=column_headers)
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
df.dtypes

party                   object
handicapped-infants     object
water-project           object
budget                  object
physician-fee-freeze    object
el-salvador-aid         object
religious-groups        object
anti-satellite-ban      object
aid-to-contras          object
mx-missile              object
immigration             object
synfuels                object
education               object
right-to-sue            object
crime                   object
duty-free               object
south-africa            object
dtype: object

In [0]:
df = df.replace({'y':1, "n":0, "?":np.NaN})
df.head()


party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
print(df.at[1, 'party'])
print(df.at[3, 'party'])
rep = df[df['party'] == 'republican']
dem = df[df['party'] == 'democrat']

republican
democrat


Predictions- 

Republicans will *not* support 'mx-missile' more than Democrats

Republicans will support the 'right to sue'

Both parties will equally support (mostly) 'water project'

In [0]:
print("2 Sample T-Test of the Mx-Missle bill. We predict greater Democrat than Republican Support")
stats.ttest_ind(dem['mx-missile'], rep['mx-missile'], nan_policy='omit')

2 Sample T-Test of the Mx-Missle bill. We predict greater Democrat than Republican Support


Ttest_indResult(statistic=16.437503268542994, pvalue=5.03079265310811e-47)

Conclusion: Due to the t-static of 16.4 and the p-value of 0, we fail to reject the null hypothesis that Republicans will support this bill more than Democrats. 

In [0]:
print("2 Sample T-Test of the Right-To-Sue. We predict greater Republican than Democrat support.")
stats.ttest_ind(rep['right-to-sue'], dem['right-to-sue'], nan_policy='omit')

2 Sample T-Test of the Right-To-Sue. We predict greater Republican than Democrat support.


Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

Conclusion: Due to the t-static value of 13.5 and the p-value of 0, we fail to reject the null hypothesis that Democrats will support this bill more than Republicans.

In [0]:
print("2 Sample T-Test of the Water-Projects bill. We predict comperable support between the parties.")
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')


2 Sample T-Test of the Water-Projects bill. We predict comperable support between the parties.


Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

Conclusion: Due to the t-static value of .089 and the p-value of .9, we fail to reject the null hypothesis. There is no significant statistical difference.

In [0]:
print(rep['water-project'].mean())
print(dem['water-project'].mean())

0.5067567567567568
0.502092050209205


## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!

In [0]:
#function for easy comparison
rep = df[df['party'] == 'republican']
dem = df[df['party'] == 'democrat']
def easy_comp(bill):
  return stats.ttest_ind(rep[bill], dem[bill], nan_policy='omit')

In [0]:
easy_comp('water-project')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [0]:
easy_comp('right-to-sue')

Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

In [0]:
easy_comp('mx-missile')

Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47)

In [0]:
def testme(bill):
  x_bar = np.mean(bill)
  s = np.std(bill, ddof=1)
  n = len(bill)
  mu = .5
  t = (x_bar - mu) / (s / np.sqrt((n)))
  return t

In [0]:
# Running the function-
testme(rep['water-project'].dropna())

0.1638576060745838

In [0]:
stats.ttest_1samp(rep['water-project'].dropna(), .5)

Ttest_1sampResult(statistic=0.16385760607458383, pvalue=0.8700683158521967)

In [0]:
# testing 2 sample functions-

def twice(bill):
  x1 = np.mean(rep[bill])
  x2 = np.mean(dem[bill])
  s1 = np.std(rep[bill].dropna(), ddof=1)
  s2 = np.std(dem[bill].dropna(), ddof=1)
  n1 = len(rep[bill].dropna())
  n2 = len(dem[bill].dropna())
  top = (x1 - x2)
  bottom = np.sqrt((s1*s1 / n1)+(s2*s2 / n2))
  t = top / bottom
  return t 


In [0]:
twice('water-project')

0.08893998898558053

Hell. Yes.
Writing code to do math is so much better than just doing math. Excellent.